ConnectionManager.php

```diff
@@ -0,0 +1,190 @@
+<?php
+include_once("iDatabase.php");
+/**
+ * Created by PhpStorm.
+ * User: gkalligeros
+ * Date: 11/18/2017
+ * Time: 3:14 PM
+ * Main class connection manager depends upon an database abstraction and a CacheManager
+ *
+ */
+class ConnectionManager
+{
+
+
+    private $db, $cacheManager;
+
+    /**
+     * ConnectionManager constructor.
+     * @param iDataBase $db
+     * @param CacheManager $cacheManager
+     */
+    public function __construct(iDataBase $db, CacheManager $cacheManager)
+    {
+        $this->db = $db;
+        $this->db->connect();
+        $this->cacheManager = $cacheManager;
+    }
+
+    /**
+     * Get requested value from cache if it is available or
+     * Generate a select statement with bindings and pass it to database for execution  then get the result store it in cache and return it
+     * @param string $tableName
+     * @param array $selectColumns
+     * @param  array $conditions
+     * @return bool
+     */
+    public function select($tableName, $selectColumns, $conditions)
+    {
+        $parsedColumns = implode(',', $selectColumns);
+        $query = "SELECT {$parsedColumns}  FROM {$tableName} WHERE ";
+        $conditionsParsed = [];
+        $bindings = [];
+        $result = $this->getCachedQueryResult($tableName, $selectColumns, $conditions);
+        if ($result != null) {
+            return $result;
+        }
+        foreach ($conditions as $condition) {
+            $placeHolderName = uniqid();
+            $conditionsParsed[] = " {$condition['column']} {$condition['operator']} :{$placeHolderName}";
+            $bindings[$placeHolderName] = $condition['value'];
+        }
+        $query .= implode(' AND ', $conditionsParsed);
+        $result = $this->db->execute($query, $bindings);
+        $this->cacheManager->remember($tableName, $selectColumns, $conditions, $result);
+        return $result;
+    }
+
+
+    /**
+     * Generate an insert statement with bindings and pass it to database for execution
+     * @param $tableName
+     * @param $data
+     * @return bool
+     */
+    public function insert($tableName, $data)
+    {
+        $columns = implode(array_keys($data[0]), ',');
+        $query = "INSERT INTO `{$tableName}` ({$columns}) VALUES ";
+        $bindings = [];
+        foreach ($data as $row) {
+            $placeHolders = [];
+            foreach ($row as $columnName => $value) {
+                $placeHolderName = uniqid();
+                $placeHolders[] = ":{$placeHolderName}";
+                $bindings[$placeHolderName] = $value;
+            }
+            $placeHolders = implode(',', $placeHolders);
+            $query .= "({$placeHolders})";
```

Comment:
@tsirigotis @gkalligeros I see that you sorting out sql injection by using placeholders and assuming that the db driver would escape the data correctly. How would you rest assure that the db driver has been implemented correctly?
Also it seems that the `$tablename` and the `$columns` are injected in the query, so these too could be vulnerable as well. Isn't  there any check that can be made as well?

Reply:
We should escape those..

Reply:
And use prepared statements in ddlb driver

Reply:
The connection manager should dispatch prepared statement to database ..or the database should handle queries as prepared statements

Reply:
Table name an column name should be checked against a white list of allowed strings

Reply:
Also we can escape all SQL keywords